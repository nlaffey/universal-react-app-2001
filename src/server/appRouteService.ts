import * as url from 'url';
import { getRouteCss, insertCss } from '../utils/css';
import getInitialPropsRecursively, { InitialPropsContext } from '../utils/getInitialPropsRecursively';
import { renderToString } from 'react-dom/server';
import { renderIndexHtmlTemplate } from './index-html-template';
import { getEntry } from '../services/contentful/contentfulService';
import { FULL_BUNDLE_URL } from '../constants/pathnames';
import { ResolveContext } from '../typings/server';
import { NextFunction, Request, Response } from 'express';
// noinspection TsLint
import UniversalRouter from 'universal-router';

type RouteHandlerContext = { port: number, universalRouter: any };

export type AppRouteHandlers = {
  handleUniversalRouterComponentRendering: (req: Request, res: Response, next: NextFunction) => Promise<void>
  handleContentfulGetRequest: (req: Request, res: Response, next: NextFunction) => Promise<void>
};

export type CreateAppRouteHandlers = (routeHandlerContext: RouteHandlerContext) => AppRouteHandlers;

/**
 *
 */
async function resolveComponentAndInitialPropsWithUniversalRouter(universalRouter: UniversalRouter, resolveContext: ResolveContext, port: number) {
  const component = await universalRouter.resolve(resolveContext);
  const initialPropsContext: InitialPropsContext = { port, resolveContext };
  const initialProps = await getInitialPropsRecursively(component, initialPropsContext);
  const resolveContextWithProps = { ...resolveContext, initialProps };
  return await universalRouter.resolve(resolveContextWithProps);
}

const createAppRouteHandlers: CreateAppRouteHandlers =
  function createAppRouteHandlers(routeHandlerContext: RouteHandlerContext): AppRouteHandlers {
    const { port, universalRouter } = routeHandlerContext;

    async function handleUniversalRouterComponentRendering(req: Request, res: Response, next: NextFunction) {
      const parsedUrl = url.parse(req.url);
      const { query, pathname } = parsedUrl;
      const context = { insertCss };
      const resolveObject: ResolveContext = { pathname, query, context };
      try {
        const componentWithProps = await resolveComponentAndInitialPropsWithUniversalRouter(universalRouter, resolveObject, port);
        const componentHtml = renderToString(componentWithProps);
        const css = getRouteCss();
        const html = renderIndexHtmlTemplate(componentHtml, FULL_BUNDLE_URL, css);
        res.send(html);
      } catch (err) {
        if (err.status !== 404) {
          console.error(err);
        }
        next();
      }
    }

    async function handleContentfulGetRequest(req: Request, res: Response) {
      const { entryId } = req.params;
      res.setHeader('Content-Type', 'application/json');
      res.setHeader('Cache-Control', 'public, max-age=31536000');
      const entryData = await getEntry(entryId);
      res.send(JSON.stringify(entryData));
    }

    return {
      handleUniversalRouterComponentRendering,
      handleContentfulGetRequest
    };
  };

export type AppRouteService = { createAppRouteHandlers: CreateAppRouteHandlers };
const appRouteService: AppRouteService = { createAppRouteHandlers };

export default appRouteService;
