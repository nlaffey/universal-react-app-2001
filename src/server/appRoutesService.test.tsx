import { AppRouteHandlers, AppRouteService } from './appRouteService';

describe('AppRouteService', () => {
  let appRouteHandlerService: AppRouteService;
  beforeAll(() => {
    jest.mock('../contentful/service.ts');
    jest.mock('../utils/getInitialProps.ts');
    jest.mock('react-dom/server');
    appRouteHandlerService = require('./appRouteService').default;
  });

  describe('createAppRouteHandlers', () => {
    test('should create route functions that match the snapshot', () => {
      const appRouteHandlers = appRouteHandlerService.createAppRouteHandlers(200, jest.fn());
      expect(appRouteHandlers).toMatchSnapshot();
    });
  });

  describe('handleContentfulGetRequest', () => {
    let appRouteHandlers: AppRouteHandlers;
    beforeEach(() => {
      appRouteHandlers = appRouteHandlerService.createAppRouteHandlers(2000, jest.fn());
    });

    test('should call contentful api and send entry json', async () => {
      const req: any = { params: { entryId: 'mockId' } };
      const res: any = { send: jest.fn(), setHeader: jest.fn() };
      const next: any = {};
      await appRouteHandlers.handleContentfulGetRequest(req, res, next);
      expect(res.send.mock.calls).toMatchSnapshot();
    });

    test('should set header information for cached json', async () => {
      const req: any = { params: { entryId: 'mockId' } };
      const res: any = { send: jest.fn(), setHeader: jest.fn() };
      const next: any = {};
      await appRouteHandlers.handleContentfulGetRequest(req, res, next);
      expect(res.setHeader.mock.calls).toMatchSnapshot();
    });
  });

  describe('handleUniversalRouterComponentRendering', () => {
    let appRouteHandlers: AppRouteHandlers;
    let mockUniversalRouter;
    beforeEach(() => {
      mockUniversalRouter = { resolve: jest.fn(() => Promise.resolve({ mockComponent: true })) };
      appRouteHandlers = appRouteHandlerService.createAppRouteHandlers(2000, mockUniversalRouter);
    });

    test('should call the universal router with the context, pathname and query', async () => {
      const req: any = { url: '/mockPath123?foo=baz' };
      const res: any = { send: jest.fn(), setHeader: jest.fn() };
      const next: any = {};
      await appRouteHandlers.handleUniversalRouterComponentRendering(req, res, next);
      expect(mockUniversalRouter.resolve.mock.calls).toMatchSnapshot();
    });

    test('should call getInitialProps method passing in the component and context info', async () => {
      const req: any = { url: '/mockPath123?foo=baz' };
      const res: any = { send: jest.fn(), setHeader: jest.fn() };
      const next: any = {};
      const { getInitialProps } = require('../utils/getInitialProps');
      getInitialProps.mockReset();
      await appRouteHandlers.handleUniversalRouterComponentRendering(req, res, next);
      expect(getInitialProps.mock.calls).toMatchSnapshot();
    });

    test('should send html with initial props populated', async (done) => {
      const req: any = { url: '/mockPath123?foo=baz' };
      const res: any = { send: jest.fn(), setHeader: jest.fn() };
      const next: any = {};
      await appRouteHandlers.handleUniversalRouterComponentRendering(req, res, next);
      // TODO: Look into why i'm needing to use a timeout here
      window.setTimeout(() => {
        expect(res.send.mock.calls).toMatchSnapshot();
        done();
      });
    });
  });
});
