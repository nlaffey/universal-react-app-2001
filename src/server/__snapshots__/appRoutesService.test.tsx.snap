// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppRouteService createAppRouteHandlers should create route functions that match the snapshot 1`] = `
Object {
  "handleContentfulGetRequest": [Function],
  "handleUniversalRouterComponentRendering": [Function],
}
`;

exports[`AppRouteService handleContentfulGetRequest should call contentful api and send entry json 1`] = `
Array [
  Array [
    "{\\"isGetEntryMock\\":true}",
  ],
]
`;

exports[`AppRouteService handleContentfulGetRequest should set header information for cached json 1`] = `
Array [
  Array [
    "Content-Type",
    "application/json",
  ],
  Array [
    "Cache-Control",
    "public, max-age=31536000",
  ],
]
`;

exports[`AppRouteService handleUniversalRouterComponentRendering should call getInitialProps method passing in the component and context info 1`] = `
Array [
  Array [
    Object {
      "mockComponent": true,
    },
    Object {
      "port": 2000,
      "resolveContext": Object {
        "context": Object {
          "insertCss": [Function],
        },
        "pathname": "/mockPath123",
        "query": "foo=baz",
      },
    },
  ],
  Array [
    Object {
      "mockComponent": true,
    },
    Object {
      "port": 2000,
      "resolveContext": Object {
        "context": Object {
          "insertCss": [Function],
        },
        "pathname": "/mockPath123",
        "query": "foo=baz",
      },
    },
  ],
]
`;

exports[`AppRouteService handleUniversalRouterComponentRendering should call the universal router with the context, pathname and query 1`] = `
Array [
  Array [
    Object {
      "context": Object {
        "insertCss": [Function],
      },
      "pathname": "/mockPath123",
      "query": "foo=baz",
    },
  ],
  Array [
    Object {
      "context": Object {
        "insertCss": [Function],
      },
      "initialProps": Object {
        "mockInitialProps": true,
      },
      "pathname": "/mockPath123",
      "query": "foo=baz",
    },
  ],
]
`;

exports[`AppRouteService handleUniversalRouterComponentRendering should send html with initial props populated 1`] = `
Array [
  Array [
    "<!doctype html>
    <html>
      <head>
        <title>Universal React App</title>
        <style type=\\"text/css\\">
        
        </style>
      </head>
      <body>
      <div id=\\"mounting-point\\">
        undefined
      </div>
      <script src=\\"/public/bundle.js\\"></script>
      </body>
    </html>",
  ],
]
`;
